cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh

```diff
@@ -974,7 +974,9 @@ void build_hierarchical(const handle_t& handle,
   rmm::mr::device_memory_resource* device_memory = nullptr;
   IdxT max_minibatch_size =
     calc_minibatch_size(n_clusters, n_rows, dim, metric, std::is_same_v<T, float>);
-  auto pool_guard = raft::get_pool_memory_resource(device_memory, max_minibatch_size * dim * 4);
+  auto pool_guard = raft::get_pool_memory_resource(
+    device_memory,
+    std::min<size_t>(size_t(dim * 4) * max_minibatch_size, 1000lu * 1000lu * 1000lu));
```

Comment:
`calc_minibatch_size` could return the `size_per_row`. It would be better to use that in place of `(dim*4)`. That way, if we actually do not need a large pool, then we will not allocate that.

Reply:
Yes, makes sense!